REGRESSION(294820@main): [ iOS Debug ] fast/page-color-sampling/basic-fixed-container-edge-sampling.html is a constant crash.

whsieh opened the pull request:
#### 834b04082696bc035bee94598aa0ed33554c03b2
<pre>
REGRESSION(294820@main): [ iOS Debug ] fast/page-color-sampling/basic-fixed-container-edge-sampling.html is a constant crash.
<a href="https://bugs.webkit.org/show_bug.cgi?id=294225">https://bugs.webkit.org/show_bug.cgi?id=294225</a>
<a href="https://rdar.apple.com/152870409">rdar://152870409</a>

Reviewed by Abrar Rahman Protyasha.

Adjust how `UIColor` / `NSColor` → CSS serialization works in WebCoreTestSupport, such that it&apos;s
robust in the case where any of the color components are either `-0.000000` or an otherwise
extremely small negative value, by using `roundAndClampToSRGBALossy` on the `CGColor`.

This sidesteps a debug assertion in test code, when running a handful of page color sampling tests.

* Source/WebCore/testing/cocoa/CocoaColorSerialization.mm:
(WebCoreTestSupport::serializationForCSS):

Canonical link: <a href="https://commits.webkit.org/296120@main">https://commits.webkit.org/296120@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/44f173315f6d98e70e3b1d7c6d03959837d111c9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107474 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27156 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17564 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112688 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/58010 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109438 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27840 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35657 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/81580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/58010 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110403 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/22048 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/96859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/61965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/21484 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/15000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/57454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/91414 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/15031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115789 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34541 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/25449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/90618 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34916 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/93115 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/90364 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/35276 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/13052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/30286 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17373 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34462 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/40004 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34208 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37563 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35869 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/testing/cocoa/CocoaColorSerialization.mm:
```diff
@@ -29,11 +29,21 @@
 String serializationForCSS(WebCore::CocoaColor *color)
 {
-    return WebCore::serializationForCSS(WebCore::colorFromCocoaColor(color));
+    return WebCore::serializationForCSS([&] -> WebCore::Color {
```
nit, just capture `color`

aprotyas commented on Source/WebCore/testing/cocoa/CocoaColorSerialization.mm:
```diff
@@ -29,11 +29,21 @@
+    return WebCore::serializationForCSS([&] -> WebCore::Color {
+        if (auto colorComponents = WebCore::roundAndClampToSRGBALossy([color CGColor]))
+            return { WTFMove(*colorComponents) };
+        return WebCore::colorFromCocoaColor(color);
```
test comment

whsieh committed:
REGRESSION(294820@main): [ iOS Debug ] fast/page-color-sampling/basic-fixed-container-edge-sampling.html is a constant crash.
https://bugs.webkit.org/show_bug.cgi?id=294225
rdar://152870409

Reviewed by Abrar Rahman Protyasha.

Adjust how `UIColor` / `NSColor` → CSS serialization works in WebCoreTestSupport, such that it's
robust in the case where any of the color components are either `-0.000000` or an otherwise
extremely small negative value, by using `roundAndClampToSRGBALossy` on the `CGColor`.

This sidesteps a debug assertion in test code, when running a handful of page color sampling tests.

* Source/WebCore/testing/cocoa/CocoaColorSerialization.mm:
(WebCoreTestSupport::serializationForCSS):

Canonical link: https://commits.webkit.org/296120@main
